File: 0001-0100/0092.Reverse_Linked_List_II.py
//0ms


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    # Helper function to reverse a linked list
    def reverseLL(self, head):
        if not head or not head.next:
            return head
        last = self.reverseLL(head.next)
        head.next.next = head
        head.next = None
        return last

    def reverseBetween(self, head, left, right):
        if left == right or not head:
            return head
        
        # Dummy node to handle edge cases (e.g., reversing from head)
        dummy = ListNode(0)
        dummy.next = head
        prev = dummy

        # Move to the node before 'left'
        for _ in range(1, left):
            prev = prev.next

        # Identify the start and end of the sublist
        start = prev.next
        end = start
        for _ in range(left, right):
            end = end.next

        # Save the rest of the list after 'right'
        rest = end.next
        end.next = None  # Isolate the sublist

        # Reverse the sublist
        reversed_sublist = self.reverseLL(start)

        # Reconnect the reversed sublist back to the main list
        prev.next = reversed_sublist
        start.next = rest

        return dummy.next

File: 0001-0100/0092.Reverse_Linked_List_II.java
//0ms


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseLL(ListNode head){
        if(head==null || head.next==null) return head;
        ListNode last = reverseLL(head.next);
        head.next.next = head;
        head.next = null;
        return last;
    }
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if(left==right || head==null) return head;
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode prev = dummy;
        for(int i=1;i<left;i++) prev = prev.next;
        ListNode start=prev.next, end=start;
        for(int i=left;i<right;i++) end=end.next;
        ListNode res = end.next;
        end.next = null;
        ListNode reversed = reverseLL(start);
        prev.next=reversed;
        start.next = res;
        return dummy.next;
    }
}
